sdk/appconfiguration/azure-spring-cloud-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/CalculatedBackoffTime.java

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.spring.cloud.config;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.TimeUnit;
+
+import com.azure.spring.cloud.config.properties.AppConfigurationProviderProperties;
+
+/**
+ * Calculates the amount of time to the next refresh, if a refresh fails.
+ */
+public final class CalculatedBackoffTime {
+
+    /**
+     * Calculates the amount of time to the next refresh, if a refresh fails. Takes current Refresh date into account
+     * for watch keys.
+     * @param state state being checked
+     * @param properties App Configuration Provider Properties
+     * @return new Refresh Date
+     */
+    public static Instant calculateBefore(State state, AppConfigurationProviderProperties properties) {
+        // The refresh interval is only updated if it is expired.
+        if (Instant.now().isAfter(state.getNextRefreshCheck())) {
+            Integer attempt = state.getRefreshAttempt();
+            state.addRefreshAttempt();
+
+            return calculate(attempt, state.getRefreshInterval(), properties);
+        }
+
+        return state.getNextRefreshCheck();
+    }
+
+    /**
+     * Calculates the amount of time to the next refresh, if a refresh fails. Takes current Refresh date into account
+     * for watch keys. Used for checking client refresh-interval only.
+     * @param nextRefreshCheck next refresh for the whole client
+     * @param attempt refresh attempt for the client
+     * @param interval the Refresh Interval
+     * @param properties App Configuration Provider Properties
+     * @return new Refresh Date
+     */
+    public static Instant calculateBefore(Instant nextRefreshCheck, Integer attempt, Duration interval,
+        AppConfigurationProviderProperties properties) {
+
+        if (interval == null) {
+            return null;
+        }
+
+        // The refresh interval is only updated if it is expired.
+        if (Instant.now().isAfter(nextRefreshCheck)) {
+            return calculate(attempt, interval.getSeconds(), properties);
+        }
+
+        return nextRefreshCheck;
+    }
+
+    private static Instant calculate(Integer attempt, long interval, AppConfigurationProviderProperties properties) {
+        int durationPeriod = Math.toIntExact(interval);
+
+        Instant now = Instant.now();
+
+        if (durationPeriod <= properties.getDefaultMinBackoff()) {
+            return now.plusSeconds(interval);
```

Comment:
The main purpose of this class is to calculate randomized exponential backoff time. That's a generic utility that we should be able to reuse whenever we need to calculate randomized exponential backoff in future (for example, geo replication work will involve similar calculations).
So I think we should just return just the calculated backoff time (eg: 1s,2s,4s,...) instead of the next refresh time.

If you agree, we can think of a more suitable method name than the one I suggested in my previous comments. How about `BackoffTimeCalculator.getNextBackoff(...)`?

Reply:
#Resolved